dlib/python/numpy.h

```diff
@@ -114,5 +108,97 @@ void get_numpy_ndarray_parts (
 
 // ----------------------------------------------------------------------------------------
 
+template <typename T, int dims>
+void get_numpy_ndarray_parts (
+    boost::python::object& obj,
+    T*& data,
+    dlib::array<T>& contig_buf,
+    long (&shape)[dims]
+)
+/*!
+    ensures
+        - extracts the pointer to the data from the given numpy ndarray.  Stores the shape
+          of the array into #shape.
+!*/
+{
+    Py_buffer pybuf;
+    if (PyObject_GetBuffer(obj.ptr(), &pybuf, PyBUF_STRIDES | PyBUF_WRITABLE ))
+        throw dlib::error("Expected writable numpy.ndarray with shape set.");
+
+    try
+    {
+        validate_numpy_array_type<T>(obj);
+
+        get_numpy_ndarray_shape(obj, shape);
+
+        if (dlib::pixel_traits<T>::num > 1 && dlib::pixel_traits<T>::num != shape[dims-1])
+            throw dlib::error("Expected numpy.ndarray with " + dlib::cast_to_string(dlib::pixel_traits<T>::num) + " channels.");
+
+        if (PyBuffer_IsContiguous(&pybuf, 'C'))
+            data = (T*)pybuf.buf;
+        else
+        {
+            contig_buf.resize(pybuf.len);
+            if (PyBuffer_ToContiguous(&contig_buf[0], &pybuf, pybuf.len, 'C'))
+                throw dlib::error("Can't copy numpy.ndarray to a contiguous buffer.");
+            data = (T*)&contig_buf[0];
```

Comment:
You shouldn't need this cast to T*.  It's better to not have casts when you don't need them.